File: frontend/src/services/categoryService.ts
// ============================================================
// CategoryService — acesso à API de categorias e propagação em massa.
// CardDetailService — acesso à API de detalhes de cartão de crédito.
// ============================================================

import type {
  Category,
  CreateCategoryInput,
  BulkCategoryUpdateInput,
  BulkCategoryUpdateResult,
  CardDetail,
  ImportCardDetailInput,
} from '@/types/category';
import type { ApiResponse } from '@/types';
import apiClient from './api';

// ---- CategoryService ----

class CategoryService {
  private readonly endpoint = '/categories';

  // Busca todas as categorias, opcionalmente filtradas por tipo.
  async getAll(type?: 'income' | 'expense'): Promise<Category[]> {
    const response = await apiClient.get<ApiResponse<Category[]>>(
      this.endpoint,
      { params: type ? { type } : undefined }
    );
    return response.data.data ?? [];
  }

  // Cria uma nova categoria personalizada.
  async create(input: CreateCategoryInput): Promise<Category> {
    const response = await apiClient.post<ApiResponse<Category>>(this.endpoint, input);
    if (!response.data.data) throw new Error('Erro ao criar categoria');
    return response.data.data;
  }

  // Remove uma categoria pelo ID.
  async delete(id: string): Promise<void> {
    await apiClient.delete(`${this.endpoint}/${id}`);
  }

  // Propaga uma nova categoria para todas as transações com título ≥50% similar.
  // Chamado automaticamente após edição inline de categoria na lista de transações.
  async bulkUpdate(input: BulkCategoryUpdateInput): Promise<BulkCategoryUpdateResult> {
    const response = await apiClient.post<BulkCategoryUpdateResult>(
      `${this.endpoint}/bulk-update`,
      input
    );
    return response.data;
  }
}

export const categoryService = new CategoryService();

// ---- CardDetailService ----

class CardDetailService {
  private readonly endpoint = '/transactions';

  // Busca os detalhes de subcategorização de uma fatura de cartão.
  // Retorna null quando nenhum detalhe foi importado ainda.
  async getByTransaction(transactionId: string): Promise<CardDetail | null> {
    const response = await apiClient.get<ApiResponse<CardDetail>>(
      `${this.endpoint}/${transactionId}/card-details`
    );
    return response.data.data ?? null;
  }

  // Importa os itens de detalhe de uma fatura.
  // O backend valida que a soma não ultrapassa o valor da fatura
  // e registra automaticamente o restante como "Outros".
  async import(transactionId: string, input: ImportCardDetailInput): Promise<CardDetail> {
    const response = await apiClient.post<ApiResponse<CardDetail>>(
      `${this.endpoint}/${transactionId}/card-details`,
      input
    );
    if (!response.data.data) throw new Error('Erro ao importar detalhes');
    return response.data.data;
  }
}

export const cardDetailService = new CardDetailService();

File: backend/services/category_service.go
package services

import (
	"context"
	"errors"

	"github.com/user/financas-api/interfaces"
	"github.com/user/financas-api/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// ---- CategoryService ----

// CategoryService implementa interfaces.CategoryService.
type CategoryService struct {
	repo interfaces.CategoryRepository
}

func NewCategoryService(repo interfaces.CategoryRepository) interfaces.CategoryService {
	return &CategoryService{repo: repo}
}

// EnsureDefaults popula a collection com categorias padrão se estiver vazia.
// Chamado no startup da aplicação via routes.Setup().
func (s *CategoryService) EnsureDefaults(ctx context.Context) error {
	return s.repo.Seed(ctx)
}

// GetAll retorna as categorias filtradas por tipo ("income", "expense" ou "").
func (s *CategoryService) GetAll(ctx context.Context, categoryType string) ([]models.Category, error) {
	return s.repo.FindAll(ctx, models.CategoryType(categoryType))
}

// Create adiciona uma nova categoria personalizada ao banco.
func (s *CategoryService) Create(ctx context.Context, input *models.CreateCategoryInput) (*models.Category, error) {
	category := &models.Category{
		Name: input.Name,
		Type: input.Type,
	}
	return s.repo.Create(ctx, category)
}

// Delete remove uma categoria pelo ID em formato string.
func (s *CategoryService) Delete(ctx context.Context, id string) error {
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return errors.New("ID de categoria inválido")
	}
	return s.repo.Delete(ctx, objectID)
}

// ---- CardDetailService ----

// CardDetailService implementa interfaces.CardDetailService.
// Valida que a soma dos itens não ultrapassa o valor da fatura e registra o restante como "Outros".
type CardDetailService struct {
	repo interfaces.CardDetailRepository
}

func NewCardDetailService(repo interfaces.CardDetailRepository) interfaces.CardDetailService {
	return &CardDetailService{repo: repo}
}

// GetByTransactionID retorna os detalhes de detalhe de uma fatura.
func (s *CardDetailService) GetByTransactionID(ctx context.Context, transactionID string) (*models.CardDetail, error) {
	objectID, err := primitive.ObjectIDFromHex(transactionID)
	if err != nil {
		return nil, errors.New("ID de transação inválido")
	}
	return s.repo.FindByTransactionID(ctx, objectID)
}

// Import valida e salva os itens de detalhe de uma fatura de cartão.
// Regras de negócio:
//  1. A soma dos itens importados não pode ultrapassar invoiceTotal.
//  2. O valor restante (invoiceTotal - soma) é registrado automaticamente como "Outros".
//  3. Se não houver restante (ou for < 0,01), nenhum item "Outros" é criado.
func (s *CardDetailService) Import(
	ctx context.Context,
	transactionID string,
	invoiceTotal float64,
	input *models.ImportCardDetailInput,
) (*models.CardDetail, error) {
	objectID, err := primitive.ObjectIDFromHex(transactionID)
	if err != nil {
		return nil, errors.New("ID de transação inválido")
	}

	// Valida e soma os itens importados
	var allocatedTotal float64
	for i, item := range input.Items {
		if item.Amount <= 0 {
			return nil, errors.New("todos os valores devem ser maiores que zero")
		}
		allocatedTotal += item.Amount
		if item.Name == "" {
			input.Items[i].Name = "Item sem nome"
		}
	}

	if allocatedTotal > invoiceTotal+0.01 {
		return nil, errors.New("a soma dos itens ultrapassa o valor da fatura")
	}

	// Calcula o restante e registra como "Outros" se necessário
	remaining := invoiceTotal - allocatedTotal
	items := input.Items

	const minRemainder = 0.01
	if remaining > minRemainder {
		items = append(items, models.CardDetailItem{
			Name:     "Outros",
			Amount:   roundTwo(remaining),
			Category: "Outros",
		})
		remaining = 0
	}

	detail := &models.CardDetail{
		TransactionID:  objectID,
		Items:          items,
		InvoiceTotal:   roundTwo(invoiceTotal),
		AllocatedTotal: roundTwo(allocatedTotal),
		Remaining:      roundTwo(remaining),
	}

	return s.repo.Upsert(ctx, detail)
}

// roundTwo arredonda um float64 para 2 casas decimais.
func roundTwo(v float64) float64 {
	return float64(int(v*100+0.5)) / 100
}
